Vorige week heb je met behulp van Keras een neuraal netwerk gebouwd voor de MNIST‑dataset.


Dit netwerk gebruikte ingebouwde activatiefuncties zoals relu en softmax.



Deze week ga je:

- Activatiefuncties zelf implementeren
- Begrijpen wat pre‑activatie, logits en outputinterpretatie zijn
- Onderzoeken wat het effect van verschillende activatiefuncties is op het gedrag van je netwerk

### Pre‑activatie expliciet maken

Normaal voert Keras deze stap automatisch uit. In deze opdracht ga je dit zichtbaar maken.

#### Pre-Activation
Gegeven één neuron met:

inputs: x
gewichten: w
bias: b


Schrijf in Python een functie:

def ***pre_activation(x, w, b)***. Deze functie moet de pre‑activatie berekenen (dus nog zonder activatiefunctie).

Test je functie met zelfgekozen waarden.
Print de uitkomst en leg in commentaar uit:

wat deze waarde betekent
waarom dit nog geen output van het neuron is

In [1]:
import math
import numpy as np

In [2]:
def pre_activation(values, weights, bias):
    return sum([values[i] * weights[i] for i, _ in enumerate(weights) ]) + bias

print(pre_activation([1, 1], [2, 2], 1))
# pre-actiavtion = (1*2 + 1*2) + 1
# Dit is nog niet de output van de neuron omdat de activator functie nog niet is gebruikt.
# Deze zijn erg belangrijk omdat deze de neuron non linear maken (waarmee het beter op de data aansluit).

5


#### Activatiefuncties zelf maken

Tot nu toe heb je default activatiefunctie implementaties gebruikt, maat laten we nu kijken naar eigen implementatie om te begrijpen hoe het nou werkt. (Mogelijk heb je enkele van de functies al gemaakt bij de CodeGrade opdrachten van deze week.)

#### ReLU zelf implementeren

Schrijf een functie:

***def relu(z):***

De functie moet werken voor:

- 1 getal
- 1 numpy array


Test je functie met:

- negatieve waarden
- nul
- positieve waarden



#### Sigmoid zelf implementeren

Zoek de formule op. en implementeer de functie:

***def sigmoid(z):***

Test de functie met:

- grote negatieve waarden
- 0
- grote positieve waarden


- Wanneer gaat de output richting 0?
- Wanneer richting 1?
- Wat betekent een output van 0.5?


#### Softmax zelf implementeren

Zoek de formule op. en implementeer de functie:

Je krijgt de formule:

***def softmax(z):*** 

Test met verschillende lijstjes.

Let er op dat de outputs optellen tot 1.

waarom softmax ook al weer nuttig is bij multi‑class classificatie?

In [3]:
values = [-100, -10, -3, -2, -1, 0, 1, 2, 3, 10, 100]

def relu(value):
    return max(0, value)

def sigmoid(value):
    return 1 / (1 + math.exp(-value))

# Output van 0.5 ligt halverwegen tussen 0 en 1. Dit komt dus overeen met een input van 0.
# Output gaat richting 0 onder de 0.5
# Output gaat richting 1 boven 0.5

softmax_values_1 = [1, 0, -1]
softmax_values_2 = [0.1, 0.1, 0.1]

def softmax(values):
    return np.exp(values) / np.sum(np.exp(values))



print([relu(v) for v in values])
print([sigmoid(v) for v in values])

print(softmax(softmax_values_1))
print(sum(softmax(softmax_values_1)))
print(softmax(softmax_values_2))
print(sum(softmax(softmax_values_2)))


[0, 0, 0, 0, 0, 0, 1, 2, 3, 10, 100]
[3.7200759760208356e-44, 4.5397868702434395e-05, 0.04742587317756678, 0.11920292202211755, 0.2689414213699951, 0.5, 0.7310585786300049, 0.8807970779778823, 0.9525741268224334, 0.9999546021312976, 1.0]
[0.66524096 0.24472847 0.09003057]
1.0
[0.33333333 0.33333333 0.33333333]
1.0


#### Je eigen Activatiefuncties gebruiken in het netwerk

Maak een Keras‑model (of gebruik die uit de vorige practicum opdracht) waarin je geen activatiefunctie opgeeft in de layers
alleen Dense(units) gebruikt


- Wat zijn de outputs van de hidden layer nu?
- Wat stellen deze waarden conceptueel voor?

Gebruik nu dan je zelfgeschreven activatiefuncties om:

De output van de hidden layer handmatig door ReLU te halen
De output van de laatste layer door:

- sigmoid (single‑class experiment)
- softmax (multi‑class experiment)

Hoe kun je dit in je mnist db gebruiken(single‑class en multi-class)? Denk er over na.

Welke heeft het beste gewerkt? Waarom?

Probeer je netwerk zo te maken dat het nog steeds op de Mystery device past.



Antwoord:
- De outputs zijn willekeurige getallen
- Deze waardes zouden samenstellingen van eerdere lagen moeten voorstellen. Momenteel slagen ze er niet in verder te komen dan een willekeurige verdeling. Met 10% accuracy.

Deze zelf gedefineerde activator fucties hebben geen toegevoegde waarde. Ze zijn al ingebouwd. Dus zorgt de extra lambda layer alleen maar voor een groter model met meer overhead.

het lijkt er op dat er weinig verschil is tussen sigmoid en softmax in voorspellingskracht.


In [4]:
# Imports
import numpy as np;
from tensorflow.keras.datasets import mnist;
import tensorflow as tf;
import random as rnd;

# 60_000 training images
# 10_000 test images

(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

I0000 00:00:1779870348.380630    1362 port.cc:153] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
I0000 00:00:1779870348.381260    1362 cudart_stub.cc:31] Could not find cuda drivers on your machine, GPU will not be used.
I0000 00:00:1779870348.431815    1362 cpu_feature_guard.cc:227] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX_VNNI FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
I0000 00:00:1779870349.476674    1362 port.cc:153] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
I0000 00:0

In [5]:
# Method Block

def OneHotEncodeDigit(label: int):
    """
    One hot encode a single one digit label
    """
    arr = np.zeros((10))
    arr[label] = 1
    return arr

def OneHotEncode(arr):
    """
    One hot encode an array of labels, like the train_images
    """
    return np.array([OneHotEncodeDigit(item) for item in arr])

def normalize_and_vectorize_image(image: np.array):
    """
    Turn a image of variable dimesions (must be square) into a vector and normalize values between 0 and 1
    """
    flat = image.flatten()

    maximum = np.max(flat)
    normalised: np.array = flat / maximum

    return normalised.astype(np.half)

def compres_image(image, new_size = 14):
    """
    Reduce the resolution of an image
    """
    originial_size = len(image)
    step = originial_size / new_size
    steps = [s * step for s in range(new_size)]
    new_image = np.zeros((new_size, new_size), dtype="uint8")
    y_i = 0
    for block_y_start in steps:
        block_y_end = block_y_start + step
        x_i = 0
        for block_x_start in steps:
            block_x_end = block_x_start + step
            block = image[int(block_y_start) : int(block_y_end), int(block_x_start): int(block_x_end)]
            pixel_color = np.average(block)
            new_image[y_i, x_i] = pixel_color
            x_i += 1
        y_i += 1
    return new_image


In [6]:
# ENCODE
test_labels_encoded = OneHotEncode(test_labels)
train_labels_encoded = OneHotEncode(train_labels)
test_images_normalised = np.array([normalize_and_vectorize_image(image) for image in test_images])
train_images_normalised = np.array([normalize_and_vectorize_image(image) for image in train_images])

In [7]:
#Shrink
test_images_normalised_small = np.array([normalize_and_vectorize_image(compres_image(image)) for image in test_images])
print("finished test imaged")
train_images_normalised_small = np.array([normalize_and_vectorize_image(compres_image(image)) for image in train_images])

finished test imaged


In [8]:
model = tf.keras.Sequential([
    tf.keras.Input(shape=(784,)),
    tf.keras.layers.Dense(140),
    tf.keras.layers.Dense(40),
    tf.keras.layers.Dense(10)
])

model.compile(optimizer='adam',
    loss='categorical_crossentropy',
    metrics=[
        "accuracy",
    ])

model.fit(train_images_normalised, train_labels_encoded, epochs=5, batch_size=128, verbose=1)

model.evaluate(test_images_normalised, test_labels_encoded)


E0000 00:00:1779870405.406854    1362 cuda_platform.cc:52] failed call to cuInit: INTERNAL: CUDA error: Failed call to cuInit: UNKNOWN ERROR (303)


Epoch 1/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 3s 5ms/step - accuracy: 0.1178 - loss: 8.0121
Epoch 2/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 3s 6ms/step - accuracy: 0.1205 - loss: 7.9102
Epoch 3/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 3s 5ms/step - accuracy: 0.1197 - loss: 8.1222
Epoch 4/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.1199 - loss: 7.4998
Epoch 5/5
469/469 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.1199 - loss: 7.4108
313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.1213 - loss: 7.1339


[7.1338701248168945, 0.12129999697208405]

In [9]:
model.predict(np.array([test_images_normalised[1]]))

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 81ms/step


array([[ 21.019545,  32.623516, -18.025179, -23.202673, -20.753426,
         -8.858298,  14.938483,   8.229282, -25.960882,  17.996723]],
      dtype=float32)

In [10]:
# Custom activation function
from keras.layers import Activation
from keras.utils import get_custom_objects

def custom_relu(x):
    return tf.maximum(0.0, x)

def custom_softmax(x):
    return tf.exp(x) / tf.reduce_sum(tf.exp(x))

def custom_sigmoid(x):
    return 1 / (1 + tf.exp(-x))

In [11]:
model = tf.keras.Sequential([
    tf.keras.Input(shape=(784,)),
    tf.keras.layers.Dense(40),
    tf.keras.layers.Lambda(custom_relu),
    tf.keras.layers.Dense(10),
    tf.keras.layers.Lambda(custom_softmax)
])


model.compile(optimizer='adam',
    loss='categorical_crossentropy',
    metrics=[
        "accuracy",
    ])



model.fit(train_images_normalised, train_labels_encoded, epochs=50, batch_size=128, verbose=1)


model.evaluate(test_images_normalised, test_labels_encoded)

Epoch 1/50
469/469 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8684 - loss: 0.4850
Epoch 2/50
469/469 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.9365 - loss: 0.2286
Epoch 3/50
469/469 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9492 - loss: 0.1820
Epoch 4/50
469/469 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.9564 - loss: 0.1543
Epoch 5/50
469/469 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.9621 - loss: 0.1343
Epoch 6/50
469/469 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9652 - loss: 0.1203
Epoch 7/50
469/469 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9685 - loss: 0.1100
Epoch 8/50
469/469 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9717 - loss: 0.0985
Epoch 9/50
469/469 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9735 - loss: 0.0903
Epoch 10/50
469/469 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.9762 - loss: 0.0825
Epoch 11/50
469/469 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.9772 - loss: 0.0771
Epoch 12/50
469/469 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step

[0.140179842710495, 0.9692999720573425]

In [13]:
model = tf.keras.Sequential([
    tf.keras.Input(shape=(784,)),
    tf.keras.layers.Dense(40),
    tf.keras.layers.Lambda(custom_relu),
    tf.keras.layers.Dense(10),
    tf.keras.layers.Lambda(custom_sigmoid)
])


model.compile(optimizer='adam',
    loss='categorical_crossentropy',
    metrics=[
        "accuracy",
    ])



model.fit(train_images_normalised, train_labels_encoded, epochs=3, batch_size=128, verbose=1)


model.evaluate(test_images_normalised, test_labels_encoded)

Epoch 1/3
469/469 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.8731 - loss: 0.5043
Epoch 2/3
469/469 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.9315 - loss: 0.2409
Epoch 3/3
469/469 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9437 - loss: 0.1970
313/313 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.9470 - loss: 0.1892


[0.18918268382549286, 0.9470000267028809]

In [ ]:
model.summary()

In [ ]:
import tracemalloc;
import time;

In [ ]:


model_small = tf.keras.Sequential([
    tf.keras.Input(shape=(196,)),
    tf.keras.layers.Dense(3, 'relu'),
    tf.keras.layers.Dense(10, 'sigmoid')
])


model_small.compile(optimizer='adam',
    loss='categorical_crossentropy',
    metrics=[
        "accuracy",
    ])



model_small.fit(train_images_normalised_small, train_labels_encoded, epochs=2, batch_size=128, verbose=1)

time.sleep(1)

tracemalloc.start()
tracemalloc.clear_traces()

model_small.evaluate(test_images_normalised_small, test_labels_encoded)

curr, peak = tracemalloc.get_traced_memory()

print(f"curr: {curr/1000_000} mb | max {peak/1000_000} mb | differance {(peak-curr)/1000_000}")

model_small.summary()

Op basis van de hierbovenstaande gegevens is het goed mogelijk dat het model past op de mystery device. Maar zonder docker container/vm om deze na te boosten blijft het lastig hier met zekerheid uitsprken over te doen. tracemalloc geeft het beeld dat veel geheugen wordt verbruikt. Ik vermoed dat dit komt doordat er onnodige informatie wordt megerekend zoals alle mnist afbeeldingen of andere modellen. De summary van het model geeft een veel geringere grote weer van slechts 7.41 kb van alle parameters. Hier moet nog overhead bij worden gerekend waaronder de compressie functie en library imports.